### PR TITLE
add AccountsIndex::scan_accounts

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
 no current way to scan all accounts visible to a bank

 #### Summary of Changes
 - [x] add AccountsIndex::scan_accounts()
 - [ ] add something in accounts_db to expose similar to accounts and bank

Fixes #

### CODE REVIEW

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -379,6 +379,37 @@ impl AccountsDB {
...
+        let mut collector = A::default();
+        let accounts_index = self.accounts_index.read().unwrap();
+        let storage = self.storage.read().unwrap();
+        accounts_index.scan_accounts(ancestors, |pubkey, (account_info, fork)| {
+            scan_func(
+                &mut collector,
+                storage
+                    .0
+                    .get(&fork)
+                    .unwrap_or(&HashMap::new())
```
Comment (rob-solana) [inline runtime/src/accounts_db.rs]: can you and_then() here?

File: runtime/src/accounts_db.rs [hunk 2]
```diff
@@ -379,6 +379,37 @@ impl AccountsDB {
...
+        A: Default,
+    {
+        let mut collector = A::default();
+        let accounts_index = self.accounts_index.read().unwrap();
+        let storage = self.storage.read().unwrap();
+        accounts_index.scan_accounts(ancestors, |pubkey, (account_info, fork)| {
+            scan_func(
+                &mut collector,
+                storage
+                    .0
```
Comment (rob-solana) [inline runtime/src/accounts_db.rs]: can you collect `storage` as the first of the tuple above, to get rid of the `.0` ?

**Comment by rob-solana** [comment]
@CriesofCarrots it might be better to land the foundation PR first, my bad.

**Comment by CriesofCarrots** [comment]
> @CriesofCarrots it might be better to land the foundation PR first, my bad.

Np, I will remove my commit so you can merge.

**Pull request merged**